src/imitation/algorithms/bc.py

```diff
@@ -145,6 +150,9 @@ def _calculate_loss(self, obs, acts) -> Tuple[th.Tensor, Dict[str, float]]:
             stats_dict: Statistics about the learning process to be logged.
 
         """
+        obs = th.as_tensor(obs, device=self.device)
```

Comment:
What happens if these were originally tensors? Do we end up taking gradients through them (so we might change the observations to be closer to the policy rather than the other way around)?

Reply:
Hm, I wasn't considering that case. I think that when an observation comes from a DataLoader, it's almost never expected to have a gradient on it. Thoughts on whether DataLoader can or ever should return a Tensor with parents, @qxcv?

Reply:
According to [the docs](https://pytorch.org/docs/stable/generated/torch.as_tensor.html), `th.as_tensor` will copy the computational graph if the Tensor can have a gradient.

Reply:
My read is it does preserve the computational graph? "otherwise a new Tensor will be returned with **computational graph retained** if data Tensor has requires_grad=True" (emphasis mine) But I'm not sure.

Reply:
The dataloader shouldn't return tensors with `requires_grad=True`, but I don't think there's much cost to `.deatch()`ing the tensor if that's a worry.

Reply:
(TBH I've never bothered doing this in the Torch code I write unless I expect it to be genuinely ambiguous to whoever is calling the function)

Reply:
I ran some scratch code and it looks like `as_tensor` as currently implemented simply returns the argument if it's a Tensor and the device doesn't change.

So in this case it's the exact same computational graph. But I'd imagine that you would have to copy the computational graph if you changed devices?

Reply:
Anyways, I'm not sure if there's something I should be commenting on or adding a `detach` to here. I don't believe that this is a case that folks will easily run into, and if they did they probably are being intentional about it.

Reply:
What use case do we have in mind for these already being tensors? My inclination is to call `detach()` unless there's actually a case where we want to take gradients through observations.

Reply:
`pytorch.util.data.DataLoader` will automatically convert np.ndarrays from its Torch Dataset argument to `Tensor`.

I'm fine with detaching, making that change now.